ISPN-5727 Stop client listener executor on shutdown

galderz opened the pull request:
* Make sure also that when a client listener is removed, the task is
  cancelled.
* Added a test helper method to detect thread leaks given a regular
  expression.

danberindei commented on client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/ClientListenerLeakTest.java:
```diff
@@ -0,0 +1,29 @@
+            eventListener.expectOnlyCreatedEvent(1, cache());
+         }
+      });
+      detectThreadLeaks(".*Client-Listener.*");
```
Won't this fail if there's another test running in parallel and using client listeners?

danberindei replied:
Failed in CI (and on my machine):
http://ci.infinispan.org/viewLog.html?buildId=30019&buildTypeId=bt9&guest=1

galderz commented:
@danberindei Fixed the issue by adding cache name to thread name and only detecting that thread.

danberindei commented:
This failure in HotRodOperationsAuthzIT.testSupervisorAddListener seems relevant:
http://ci.infinispan.org/viewLog.html?buildId=30078&buildTypeId=bt9&guest=1

danberindei commented:
The `HotRodOperationsAuthzIT.testSupervisorAddListener` NPE is still present in the latest build.

danberindei commented:
Personally I'm not sure the thread pool is needed at all - I would expect most listeners are installed for long periods of time, so there should be no problem with every listener creating its own thread.

galderz commented:
@danberindei Should be fixed now. Removed the cancelling of individual threads since that's not really needed (fixes the failing unit test too). The important thing is making sure the thread pool is shutdown. I added a little await for termination in the end, for more predictive unit testing.

danberindei commented:
Integrated, thanks Galder!
